flang/docs/tutorials/addingIntrinsics.md

```diff
@@ -0,0 +1,383 @@
+runtime library versioning. Here, it only matters to know it must be used in all function declaration of
+the runtime API.
+
+Trim runtime takes a descriptor for the result that it will allocate, and a constant descriptor for the string argument.
+On top of that, it takes optional source file and line arguments, that are here to print source location if the runtime
+was to reach a critical error (e.g. if the provided descriptor is actually not describing a character scalar).
+
+It is rather simple and follows the standard definition of Trim in the argument naming. In general, it is best to stick
+to the intrinsic interface in terms of argument names and order if possible.
+
+The actual implementation is in `runtime/character.cpp` and is not copied here. Have a look at it. A notable point is that it
+is not using `std::string` anywhere, because that would bring the C++ runtime as a dependency.
+
+It is using some of the `Descriptor` member function such as `string.OffsetElement<>`,
+`result.Establish()`, `result.Allocate()`, or `string.ElementBytes()`. It is worth having a look at
+`flang/runtime/descriptor.h` to understand more about the descriptor format and the tools available for it.
+
+The error handling in the runtime is done with the `Terminator` class. See the `terminator.Crash()`
+and `RUNTIME_CHECK` macro. `terminator.Crash` causes an unconditional crash, while `RUNTIME_CHECK`
+will cause a crash if the boolean provided to it is not true. The runtime should always crash rather
+than reach undefined states.
+
+Last notable point is that it is using LEN_TRIM implementation rather than to re-implement it locally.
+It is always a good idea to share part of implementations that are similar between related intrinsics.
+It is OK to define internal helper functions, classes or to use templates to achieve this.
+
+
+### TRIM Lowering
+
+In the table in `lib/Lower/IntrinsicCall.cpp`, there is a line:
+```
+{"trim", &I::genTrim, {{{"string", asAddr}}}, /*isElemental*/ false}
+```
+
+It defines that when presented with a call to an intrinsic named "trim",
+that is not elemental, lowering must lower its `string` argument in memory, and pass this lowered
+argument to a function called `genTrim` that implements the actual lowering of Trim.
+
+The `genTrim` function looks like:
+```
+// TRIM
+fir::ExtendedValue
+IntrinsicLibrary::genTrim(mlir::Type resultType,
+                          llvm::ArrayRef<fir::ExtendedValue> args) {
+   // Have a look at the implementation in lib/Lower/IntrinsicCall.cpp directly
+}
+```
+
+It provides the result type and the lowered arguments as inputs, and expect the evaluated
+result to be returned. Given the argument and result types are scalar characters, the related
+`fir::Extended` must be `fir::CharBoxValue`, which is simply a container over two SSA values: one
+for the address, one for the length.
+
+The choice here was to implement Trim with the help of runtime. So what the code is preparing the arguments
```

Comment:
The phrase "So what the code is preparing the arguments" confuses me.  I'm not sure how to correct it, but it seems wrong.

Reply:
Thanks, changed to "So this code is only preparing the arguments according to the runtime interface".